Support migration windows for BIND9 (#1428)
openstack-k8s-operators/data-plane-adoption: docs_user/modules/proc_maintaining-a-migration-window-with-legacy-bind9-servers.adoc

:_mod-docs-content-type: PROCEDURE
[id="maintaining-a-migration-window-with-legacy-bind9-servers_{context}"]

= Maintaining a migration window with legacy BIND9 servers

[role="_abstract"]

> A complete immediate switchover of all DNS servers to {rhos_long} requires all DNS clients to be updated with the new addresses of the DNS servers. This may be DNS registry entries, `glue records` in other DNS zone tables or configuration in OpenStack servers. In deployments where this cannot be done immediately, this procedure configures the existing DNS servers to be used and updated until all references to them can be migrated to the {rhos_long} DNS services.

Reviewer: ```suggestion
Transition to {rhos_long} DNS services gradually when a complete, immediate switchover isn't possible. This keeps existing DNS servers active and updated while you migrate external references, registry entries, and client configurations over time.
```

> If your {rhos_prev_long} deployment uses BIND9 servers that should continue serving DNS during adoption, follow this procedure after completing the standard {dns_service} adoption. The designate worker sends RNDC commands and DNS NOTIFY messages to the legacy servers, which pull zone data from the mDNS service through AXFR on port 5354.

Reviewer: ```suggestion
If your {rhos_prev_long} deployment uses BIND9 servers that need to stay active during adoption, follow these steps after completing the standard {dns_service} adoption. The Designate worker sends RNDC commands and DNS NOTIFY messages to your legacy servers, which then fetch zone updates from the mDNS service via AXFR on port 5354.
```

      RNDC_KEYNAME=$(${!SSH_CMD} sudo grep '^key' /var/lib/config-data/ansible-generated/designate/etc/named/rndc.conf | awk '{gsub(/"/, "", $2); print $2}' | cut -f 1)
      RNDC_ALGORITHM=$(${!SSH_CMD} sudo grep algorithm /var/lib/config-data/ansible-generated/designate/etc/named/rndc.conf | awk '{gsub(/;/, "", $2); print $2}')
      CONTROLLER_IP=$(${!SSH_CMD} sudo grep 'internal_api\"' /etc/puppet/hieradata/net_ip_map.json | awk '{gsub(/"/, "", $2); gsub(/,/, "", $2); print $2}')
      cat << BINDATACR >> "$binddata_file"
    - name: controller-$i
      address: $CONTROLLER_IP
      rndcsecret: $RNDC_SECRET
      rndckeyname: $RNDC_KEYNAME
      rndcalgorithm: $RNDC_ALGORITHM
BINDATACR
    fi
  done
$ if [ -s "$binddata_file" ]; then
    external_binds_secret_file=/tmp/external_binds_secret.yaml
    cat << EOSECRETC > "$external_binds_secret_file"
apiVersion: v1
kind: Secret
metadata:
  name: designate-external-binds
  namespace: openstack
type: Opaque
stringData:
  default: |
$(cat "$binddata_file")
EOSECRETC
    oc apply -f "$external_binds_secret_file"
  fi
----
+

> where:

Reviewer: ```suggestion
where:

```

>  `CONTROLLER1_SSH`, `CONTROLLER2_SSH`, and `CONTROLLER3_SSH`::

Reviewer: ```suggestion
`CONTROLLER1_SSH`, `CONTROLLER2_SSH`, and `CONTROLLER3_SSH`::
```

      ${!SSH_CMD} "sed -i 's:allow.notify.*};:allow-notify { $INTERNALAPI_CIDR; };:' /var/lib/config-data/ansible-generated/designate/etc/named/options.conf"
      ${!SSH_CMD} "sed -i '/allow.transfer:/d' /var/lib/config-data/ansible-generated/designate/etc/named/options.conf"
      ${!SSH_CMD} "sed -i '/^options {/,/^};/{
    /^};/ i\\\tallow-transfer { '"$INTERNALAPI_CIDR"'; };
  }' /var/lib/config-data/ansible-generated/designate/etc/named/options.conf"
  done
----
+
where:

`CONTROLLER_IP`::
Specifies the `internalapi` IP address of the legacy Controller. You can be retrieve this from `/etc/puppet/hieradata/net_ip_map.json` on each Controller.

`internalapi_subnet`::
Specifies the `internalapi` subnet in CIDR notation, for example `172.17.0.0/24`.
This is a CIDR range represents the network from which RHOSO designate worker
pods send requests.

+
[NOTE]
====
The `sed` commands in the shell script example are directed at default deployments. If customizations are made to the deployment, either modify these commands or perform the modifications directly.
====

> +

Reviewer: ```suggestion
```

> Reload the BIND9 configuration:

Reviewer: ```suggestion
. Reload the BIND9 configuration:
```

+
----
$ systemctl restart designate-designatebackendbind9
----
+
[NOTE]
====
You can edit the BIND9 configuration directly because the legacy {OpenStackPreviousInstaller} control plane is being decommissioned and no longer manages these files.
====

. Open the firewall on each legacy Controller to allow RNDC and DNS traffic from the {rhos_long} pods:
+
[subs="+quotes"]
----
$ firewall-cmd --permanent --add-rich-rule='rule family="ipv4" source address="<internalapi_subnet>" port port="953" protocol="tcp" accept'
$ firewall-cmd --permanent --add-rich-rule='rule family="ipv4" source address="<internalapi_subnet>" port port="53" protocol="tcp" accept'
$ firewall-cmd --permanent --add-rich-rule='rule family="ipv4" source address="<internalapi_subnet>" port port="53" protocol="udp" accept'
$ firewall-cmd --reload
----

. Create a LoadBalancer service to expose the mDNS service on the internalapi
  network so that the legacy BIND9 servers can perform AXFR zone transfers:

>   +

Reviewer: ```suggestion
+
```
